mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp

```diff
@@ -0,0 +1,248 @@
+            const auto pPlayer = C_MomentumPlayer::GetLocalMomPlayer();
+            Vector velocity = pPlayer->GetAbsVelocity();
+            const auto pEntData = pPlayer->GetCurrentUIEntData();
+            if (!pEntData)
+                break;
+
+            m_flCurrentValue =
+                (velocity.LengthSqr() / 2.0f + gravity * (pPlayer->GetLocalOrigin().z - pEntData->m_flLastJumpZPos)) /
+                gravity;
+        }
+        break;
+    case SPEEDOMETER_UNITS_UPS:
+    default:
+        break;
+    }
+}
+
+void SpeedometerLabel::Colorize()
+{
+    switch (m_eColorizeType)
+    {
+    case SPEEDOMETER_COLORIZE_NONE:
+    {
+        SetFgColor(m_NormalColor);
+    } break;
+    case SPEEDOMETER_COLORIZE_RANGE:
+    {
+        int roundedCurrentValue = RoundFloatToInt(m_flCurrentValue);
+        Color newColor = m_NormalColor;
+        if (m_pRangeList)
+        {
+            for (auto i = 0; i < m_pRangeList->Count(); i++)
+            {
+                Range_t range = (*m_pRangeList)[i];
+                if (roundedCurrentValue >= range.min && roundedCurrentValue <= range.max)
+                {
+                    newColor = range.color;
+                    break;
+                }
+            }
+        }
+        SetFgColor(newColor);
+    }
+    break;
+    case SPEEDOMETER_COLORIZE_COMPARISON:
+    {
+        if (m_eUnitType == SPEEDOMETER_UNITS_ENERGY && m_bSupportsEnergyUnits)
+        {
+            m_flDiff = m_flCurrentValue - m_flPastValue;
+        }
+        else
+        {
+            m_flDiff = fabs(m_flCurrentValue) - fabs(m_flPastValue);
+        }
+        SetFgColor(MomUtil::GetColorFromVariation(m_flDiff, COLORIZE_DEADZONE, m_NormalColor, m_IncreaseColor,
+                                                  m_DecreaseColor));
+    }
+    break;
+    case SPEEDOMETER_COLORIZE_COMPARISON_SEPARATE:
+    {
+        SetFgColor(m_NormalColor);
+        if (m_bDrawComparison)
+        {
+            if (!m_bCustomDiff) // calculate diff unless explicity told not to
+            {
+                m_flDiff = fabs(m_flCurrentValue) - fabs(m_flPastValue);
+            }
+            m_bCustomDiff = false;
+
+            Color compareColor;
+            g_pMOMRunCompare->GetDiffColor(m_flDiff, &compareColor, true);
+
+            char diffChar = m_flDiff > 0.0f ? '+' : '-';
+            char szText[BUFSIZELOCL];
+            Q_snprintf(szText, BUFSIZELOCL, " (%c %i)", diffChar, RoundFloatToInt(fabs(m_flDiff)));
+
+            m_pComparisonLabel->SetText(szText);
+            m_pComparisonLabel->SetFgColor(compareColor);
+        }
+    }
+    break;
+    default:
+        break;
+    }
+}
+
+bool SpeedometerLabel::StartFadeout()
+{
+    Assert(HasFadeOutAnimation());
+    g_pClientMode->GetViewportAnimationController()->StopAnimationSequence(GetParent(), m_pSzAnimationName);
+    *m_pFlAlpha = 255.0f;
+    return g_pClientMode->GetViewportAnimationController()->StartAnimationSequence(m_pSzAnimationName);
```

Comment:
I've been thinking about this a lot, and this is actually an underlying issue with animations. To run one from HudAnimations.txt requires the name of the panel so that it knows what to do, but that seems silly, because the only other alternative is to *directly run animations on a panel*. I'm thinking there can be a solution in AnimationController that uses a predefined animation, but runs it for the given panel, like:

`StartAnimationSequenceForPanel( Panel *pPanel, const char *pSequenceName )`

And this would just basically run the sequence in the context of the panel calling it. In this case it would be our speedometer labels, so then the sequence would define the thing it's animating (the panel's alpha, for example) and have it animate the property it should for that control.

Unfortunately this means the animations would have to be very specific to specific panels and contexts, but does allow user customization and easy tweaking in the long run.

Reply:
That would be very useful here! I assume not in this PR though :p

Reply:
Yeah I'll tackle this in my ramp speeds branch.